pennylane_orquestra/orquestra_device.py

```diff
@@ -382,3 +387,254 @@ def qubit_operator_string(self, observable):
+        for idx in identity_indices:
+            results.insert(idx, 1)
+
+        res = self._asarray(results)
+        return res
+
+    def single_step_results(self, workflow_id):
+        """Extracts the results of a single step obtained for a workflow.
+
+        This method assumes that the workflow had a single step and that the
+        structure of the result corresponds to results sent by Orquestra API
+        v1.0.0.
+
+        Args:
+            workflow_id (str): the ID of the workflow to extract results for
+
+        Returns:
+            results (list): a list of workflow results
+        """
+        data = loop_until_finished(workflow_id, timeout=self._timeout)
+        try:
+            step_result = [v for k, v in data.items()][0]
+            results = step_result["expval"]["list"]
+        except (IndexError, KeyError, TypeError, AttributeError) as e:
+            current_status = workflow_details(workflow_id)
+            raise ValueError(
+                f"Unexpected result format for workflow {workflow_id}.\n "
+                "{''.join(current_status)}"
+            )
+        return results
+
+    @staticmethod
+    def insert_identity_res_batch(results, empty_obs_list, identity_indices):
+        """An auxiliary function for inserting values which were not computed
+        using workflows into batch results.
+
+        Computations involving the identity observable are given by theoretical
+        values rather than as part of a workflow. Therefore, such values need
+        to be inserted into the results later.
+
+        Args:
+            results (list): workflow results of the batched execution
+            empty_obs_list (list): list of indices where every observable was
+                identity
+            identity_indices (dict): maps the index of a sublist to the
+                the list of indices where the observable was an identity
+
+        Returns:
+            list: list of results
+        """
+        # Insert the lists needed for only identity results
+        for idx in empty_obs_list:
+            results.insert(idx, [])
+
+        # Insert further identity results
+        for list_idx in identity_indices.keys():
+            for iden_idx in identity_indices[list_idx]:
+                results[list_idx].insert(iden_idx, 1)
+
+        return results
+
+    def batch_execute(self, circuits, **kwargs):
+        results = []
+        idx = 0
+        file_prefix = f"{str(uuid.uuid4())}"
+
+        # Iterating through the circuits based on the allowed number of
+        # circuits per workflow
+        while idx < len(circuits):
+            end_idx = idx + self._batch_size
+            batch = circuits[idx:end_idx]
+            file_id = f"{file_prefix}-{str(idx)}"
+
+            res = self._batch_execute(batch, file_id, **kwargs)
+
+            results.extend(res)
+            idx += self._batch_size
+
+        return results
+
+    def _batch_execute(self, circuits, file_id, **kwargs):
```

Comment:
maybe it would be good to give this a more descriptive name, so that it is more easily distinguished from `batch_execute`? Maybe `multi_step_execute`?

Reply:
Good idea, changed! Thanks